src/rust/engine/ui/src/console_ui.rs

```diff
@@ -198,40 +109,299 @@ impl ConsoleUI {
+            tasks_to_display.swap_remove(&task);
+          }
+        }
+
+        let now = SystemTime::now();
+        for (n, pbar) in indicatif.bars.iter().enumerate() {
+          let maybe_label = tasks_to_display
+            .get_index(n)
+            .map(|(_, (label, start_time))| {
+              let duration_label = match now.duration_since(*start_time).ok() {
+                None => "(Waiting)".to_string(),
+                Some(duration) => format_workunit_duration_ms!((duration).as_millis()).to_string()
+              };
+              format!("{} {}", duration_label, label)
+            });
+
+          match maybe_label {
+            Some(label) => pbar.set_message(label),
+            None => pbar.set_message(""),
+          }
+        }
+      },
+      Instance::Prodash(prodash) => {
+        let tasks_to_display = &mut prodash.tasks_to_display;
+
+        // Finish any items that are no longer relevant.
+        for span_id in tasks_to_display.keys().cloned().collect::<Vec<_>>() {
+          if heavy_hitters.contains_key(&span_id) {
+            // NB `inc_by` moves the "worms" to help show continual progress
+            tasks_to_display.get_mut(&span_id).unwrap().inc();
+            continue;
+          }
+          // Drop the item to cause it to be removed from the Tree.
+          let _ = tasks_to_display.remove(&span_id).unwrap();
+        }
+
+        // Start any new items.
+        for (span_id, (description, start_time)) in heavy_hitters {
+          if tasks_to_display.contains_key(&span_id) {
+            // We're already rendering this item, and our dynamic `unit` instance will continue to
+            // handle rendering elapsed time for it.
+            continue;
+          }
+
+          // Else, it's new.
+          // NB We allow a 8 char "buffer" to allow for timing and spaces
+          let max_len = (prodash.terminal_dimensions.0 as usize) - 8;
+          let description: String = if description.len() < max_len {
+            description.to_string()
+          } else {
+            description
+              .chars()
+              .take(max_len - 3)
+              .chain(std::iter::repeat('.').take(3))
+              .collect()
+          };
+          let mut item = prodash.tree.add_child(description);
+          item.init(
+            None,
+            Some(prodash::unit::dynamic(MillisAsFloatingPointSecs)),
+          );
+          item.set(MillisAsFloatingPointSecs::start_time_to_step(&start_time));
+          tasks_to_display.insert(*span_id, item);
+        }
+      },
+    };
+  }
+
+  pub async fn teardown(self) {
+    match self {
+      Instance::Indicatif(indicatif) => {
+
+        // When the MultiProgress completes, the Term(Destination) is dropped, which will restore
+        // direct access to the Console.
+        indicatif
+          .multi_progress_task
+          .map_err(|e| format!("Failed to render UI: {}", e))
+          .boxed();
```

Comment:
```suggestion
        indicatif
          .multi_progress_task
          .map_err(|e| format!("Failed to render UI: {}", e))
          .await
```

I think that this needs to `await` the `multi_progress_task`, but I'm surprised that you're not getting an error for an unused/unawaited future here, so I might be wrong.

Reply:
Adding the `await` leads to compiler warnings/errors which I don't quite understand.

I'm going to attempt a refactor to have both cases here share some code, so this might get unified with `prodash`. We'll see.

Reply:
OK yeah `await`ing causes some kind of deadlock or equivalent "nothing is happening" scenario :unamused:

Reply:
This blocks landing, because otherwise the UI will not necessarily be torn down properly before we move on to doing other things, which could trigger race conditions. If this were prodash, then that might be ok... but can't land that for the default.